src/mbgl/util/mapbox.cpp

```diff
@@ -21,188 +23,148 @@ bool isMapboxURL(const std::string& url) {
 }
 
-std::string normalizeGlyphsURL(const std::string& baseURL, const std::string& url, const std::string& accessToken) {
-    if (!isMapboxURL(url)) {
-        return url;
+std::string normalizeGlyphsURL(const std::string& baseURL,
+                               const std::string& str,
+                               const std::string& accessToken) {
+    if (!isMapboxURL(str)) {
+        return str;
     }
 
-    const auto pathname = getMapboxURLPathname(url);
-    if (pathname.size() < 4) {
-      Log::Error(Event::ParseStyle, "Invalid glyph URL");
-      return url;
+    const URL url(str);
+    if (!equals(str, url.domain, "fonts")) {
+        Log::Error(Event::ParseStyle, "Invalid glyph URL");
+        return str;
     }
 
-    const auto& user = pathname[1];
-    const auto& fontstack = pathname[2];
-    const auto& range = pathname[3];
-
-    return baseURL + "/fonts/v1/" + user + "/" + fontstack + "/" + range + "?access_token=" + accessToken;
+    const auto tpl = baseURL + "/fonts/v1/{path}?access_token=" + accessToken;
+    return transformURL(tpl, str, url);
 }
 
-std::string normalizeTileURL(const std::string& baseURL, const std::string& url, const std::string& accessToken) {
-    if (!isMapboxURL(url)) {
-        return url;
+std::string normalizeTileURL(const std::string& baseURL,
+                             const std::string& str,
+                             const std::string& accessToken) {
+    if (!isMapboxURL(str)) {
+        return str;
     }
 
-    auto query = normalizeQuery(url);
-    return baseURL + "/v4/" + url.substr(sizeof("mapbox://tiles/") - 1, query.second - sizeof("mapbox://tiles/") + 1) + "?access_token=" + accessToken + normalizeQuery(url).first;
-}
-
-std::string canonicalizeTileURL(const std::string& url, SourceType type, uint16_t tileSize) {
-    auto tilesetStartIdx = url.find("/v4/");
-    if (tilesetStartIdx == std::string::npos) {
-        return url;
+    const URL url(str);
+    if (!equals(str, url.domain, "tiles")) {
+        Log::Error(Event::ParseStyle, "Invalid tile URL");
+        return str;
     }
 
-    tilesetStartIdx += sizeof("/v4/") - 1;
-
-    const auto tilesetEndIdx = url.find("/", tilesetStartIdx);
-    if (tilesetEndIdx == std::string::npos) {
-        return url;
-    }
-
-    auto queryIdx = url.rfind("?");
-    if (queryIdx == std::string::npos) {
-        queryIdx = url.length();
-    }
-
-    auto basenameIdx = url.rfind("/", queryIdx);
-    if (basenameIdx == std::string::npos || basenameIdx == queryIdx - 1) {
-        basenameIdx = url.length();
-    } else {
-        basenameIdx += 1;
-    }
+    const auto tpl = baseURL + "/v4/{path}?access_token=" + accessToken;
+    return transformURL(tpl, str, url);
+}
 
-    const auto extensionIdx = url.find(".", basenameIdx);
-    if (extensionIdx == std::string::npos || extensionIdx == queryIdx - 1) {
-        return url;
+std::string
+canonicalizeTileURL(const std::string& str, const SourceType type, const uint16_t tileSize) {
```

Comment:
The prior version of `canonicalizeTileURL` wasn't restricted to `*.mapbox.com`, right? It definitely needs to continue to support `*.mapbox.cn` (we should add a test for this) and should likely continue to support arbitrary/staging hostnames as well. It's only used in situations where the source URL is `mapbox://...`, so I don't think enforcing a particular hostname is necessary.